automation/orchestrator.go

```diff
@@ -191,41 +209,91 @@ func (o *Orchestrator) Shutdown() {
 	o.cancel()
 }
 
+// advanceTrigger may admit log errors
+func (o *Orchestrator) advanceTrigger(wf *workflow.Workflow, triggerID string) *workflow.Workflow {
+	for i, triggerOpt := range wf.Triggers {
+		trigType := triggerOpt["type"].(string)
+		listener, ok := o.listeners[trigType]
+		if !ok {
+			log.Debugw("advanceTrigger: listener not found", "type", trigType)
+			return wf
+		}
+		trig, err := listener.ConstructTrigger(triggerOpt)
+		if err != nil {
+			log.Debugw("advanceTrigger: error constructing trigger", "error", err)
+			return wf
+		}
+		if trig.ID() == triggerID {
+			trig.Advance()
+			w := wf.Copy()
+			w.Triggers[i] = trig.ToMap()
+			return w
+		}
+	}
+	return wf
+}
+
 // handleTrigger calls `RunWorkflow` when an `event.ETWorkflowTrigger` event is fired
 // it expects the payload for the `event.ETWorkflowTrigger` to be a workflow.ID
 // represented as a string
 func (o *Orchestrator) handleTrigger(ctx context.Context, e event.Event) error {
 	if e.Type == event.ETWorkflowTrigger {
-		wid, ok := e.Payload.(string)
+		wtp, ok := e.Payload.(*event.WorkflowTriggerEvent)
 		if !ok {
-			return fmt.Errorf("handleTrigger: expected event.Payload to be a string: %v", e.Payload)
+			return fmt.Errorf("handleTrigger: expected event.Payload to be an `event.WorkflowTriggerEvent`: %v", e.Payload)
 		}
 		go func() {
-			if err := o.RunWorkflow(ctx, workflow.ID(wid)); err != nil {
-				log.Errorf("%s", err)
+			wf, err := o.GetWorkflow(workflow.ID(wtp.WorkflowID))
+			if err != nil {
+				log.Debugw("handleTrigger: error fetching workflow", "id", wtp.WorkflowID, "err", err)
+				return
+			}
+			wf = o.advanceTrigger(wf, wtp.TriggerID)
+			wf, err = o.SaveWorkflow(wf)
+			if err != nil {
+				log.Debugw("handleTrigger: error saving workflow", "id", wtp.WorkflowID, "err", err)
+			}
+			runID := run.NewID()
+			if err := o.runWorkflow(ctx, wf, runID); err != nil {
+				log.Debugw("handleTrigger: error running workflow", "err", err)
```

Comment:
potential bug here, where a user updates a workflow as we are prepping for a run